An example of how to use Stencils

dmitrikozlov opened the pull request:
The example shows how to use Stencils to avoid repetitive compilations of the same scripts. It addresses issue #63 and demonstrates API such as:

- [RefPtr](https://searchfox.org/mozilla-esr91/rev/f3f439e007bdd4b5b1c2ba05ca706b68563413b2/mfbt/RefPtr.h#56) class, which represents reference-counted pointer, used for JS::Stencil
- [JS::SourceText](https://searchfox.org/mozilla-esr91/rev/f3f439e007bdd4b5b1c2ba05ca706b68563413b2/js/public/SourceText.h#79) class, which holds the source text
- [JS::CompileOptions](https://searchfox.org/mozilla-esr91/rev/f3f439e007bdd4b5b1c2ba05ca706b68563413b2/js/public/CompileOptions.h#290-297) class, which represents the options for compilation
- [JS::CompileGlobalScriptToStencil](https://searchfox.org/mozilla-esr91/rev/f3f439e007bdd4b5b1c2ba05ca706b68563413b2/js/public/experimental/JSStencil.h#48-59) function, which compiles given source text into JS::Stencil
- [JS::InstantiateGlobalStencil](https://searchfox.org/mozilla-esr91/rev/f3f439e007bdd4b5b1c2ba05ca706b68563413b2/js/public/experimental/JSStencil.h#80-83) function, which converts JS::Stencil into JSScript structure (JSScript is GC-structure that represents either top-level script or function script)

dmitrikozlov committed:
Adding an example of how to use Stencils

arai-a commented on examples/stencils.cpp:
```diff
@@ -0,0 +1,237 @@
+
+  JS::CompileOptions opts(cx);
+  opts.setFileAndLine(filename, linenumber);
+  opts.setNonSyntacticScope(true);
```
This `setNonSyntacticScope` call looks unnecessary.
Non-syntactic scope flag is necessary when you want to pass environment chain when executing.
the flag affects how the non-local variable works, and if it's true, the non-local variable gets slightly de-optimized, in order to access the variable in the environment chain.

dmitrikozlov replied:
My other code crashed without the call, and I left it here because I didn't understand what it did. I removed the call.

arai-a commented on examples/stencils.cpp:
```diff
@@ -0,0 +1,237 @@
+    return nullptr;
+  }
+
+  JS::CompilationStorage compileStorage;
```
`JS::CompilationStorage` is removed in [bug 1857536](https://bugzilla.mozilla.org/show_bug.cgi?id=1857536).
If this branch is going to target the next ESR (128), or 125+, this variable and the parameter to `JS::CompileGlobalScriptToStencil` needs to be removed.

but if this is targeting 124 for now, I think the current code is fine.

arai-a replied:
actually, let's remove this to make the code compatible with 125+

dmitrikozlov replied:
Done

arai-a commented on examples/README.md:
```diff
@@ -57,3 +57,7 @@ in the `meson.build` file, and add a description of it to this
   Use this in cases where defining properties and methods in your class
   upfront might be slow.
 - **modules.cpp** - Example of how to load ES Module sources.
+- **worker.cpp** - Example of how to use SpiderMonkey in multiple threads.
```
Thank you for updating the documentation!

arai-a commented on examples/stencils.cpp:
```diff
@@ -0,0 +1,237 @@
+// compilation errors, otherwise it does no error handling and simply exits
+// if something goes wrong.
+//
+// To reuse Stencils in multiple threads, you must create a JS::FrontendContext
+// in each thread that compiles Javascript.
```
Just to be clear, you can pass `JS::FrontendContext` between threads, with the following restriction:
  * `JS::SetNativeStackQuota` must be called in the thread that is going to use the frontend context
  * the frontend context shouldn't be used from 2 threads concurrently

but of course this comment sounds good.

dmitrikozlov replied:
I changed the comments.

arai-a reviewed (approved):
Thank you for adding example for Stencil!
I'm sorry for the long delay.

dmitrikozlov committed:
Stencils, adressed feedbak.
Removed setNonSyntacticScope and JS::CompilationStorage. Set dependency('mozjs-124a1')

dmitrikozlov commented:
I've completed the changes based on the feedback. Thanks.

arai-a reviewed (approved):
Excellent! Thank you!

dmitrikozlov commented:
Do I need to do something else?

arai-a commented:
no, I've asked for merge.
Will be merged shortly.

dmitrikozlov commented:
Thank you for your help!
